cita-network/src/p2p_protocol/transfer.rs

```diff
@@ -0,0 +1,134 @@
+// CITA
+// Copyright 2016-2019 Cryptape Technologies LLC.
+
+// This program is free software: you can redistribute it
+// and/or modify it under the terms of the GNU General Public
+// License as published by the Free Software Foundation,
+// either version 3 of the License, or (at your option) any
+// later version.
+
+// This program is distributed in the hope that it will be
+// useful, but WITHOUT ANY WARRANTY; without even the implied
+// warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+// PURPOSE. See the GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::cita_protocol::network_message_to_pubsub_message;
+use crate::network::{NetworkClient, RemoteMessage};
+use crate::node_manager::{AddConnectedKeyReq, InitMsg, NetworkInitReq, NodesManagerClient};
+use bytes::BytesMut;
+use libproto::{Message as ProtoMessage, TryFrom, TryInto};
+use logger::{info, warn};
+use tentacle::{
+    context::{ServiceContext, SessionContext},
+    traits::{ProtocolMeta, ServiceProtocol},
+    ProtocolId, SessionId,
+};
+use tokio::codec::length_delimited::LengthDelimitedCodec;
+
+pub struct TransferProtocolMeta {
+    id: ProtocolId,
+    network_client: NetworkClient,
+    nodes_mgr_client: NodesManagerClient,
+}
+
+impl TransferProtocolMeta {
+    pub fn new(
+        id: ProtocolId,
+        network_client: NetworkClient,
+        nodes_mgr_client: NodesManagerClient,
+    ) -> Self {
+        TransferProtocolMeta {
+            id,
+            network_client,
+            nodes_mgr_client,
+        }
+    }
+}
+
+impl ProtocolMeta<LengthDelimitedCodec> for TransferProtocolMeta {
+    fn id(&self) -> ProtocolId {
+        self.id
+    }
+    fn codec(&self) -> LengthDelimitedCodec {
+        LengthDelimitedCodec::new()
+    }
+    fn service_handle(&self) -> Option<Box<dyn ServiceProtocol + Send + 'static>> {
+        let handle = Box::new(TransferProtocol {
+            proto_id: self.id,
+            connected_session_ids: Vec::default(),
+            network_client: self.network_client.clone(),
+            nodes_mgr_client: self.nodes_mgr_client.clone(),
+        });
+        Some(handle)
+    }
+}
+
+struct TransferProtocol {
+    proto_id: ProtocolId,
+    connected_session_ids: Vec<SessionId>,
+    network_client: NetworkClient,
+    nodes_mgr_client: NodesManagerClient,
+}
+
+impl ServiceProtocol for TransferProtocol {
+    fn init(&mut self, _control: &mut ServiceContext) {}
+
```

Comment:
This `init()` needed for test ?

Reply:
This is a function needed by the trait `ServiceProtocol`, and I don't want to do something in it.